### PR TITLE
New NX-OS Upgrade Execution and State Module

### PR DESCRIPTION
**Pull request by mikewiebe**
### What does this PR do?

This PR adds a new execution and state module to upgrade/downgrade the software on NX-OS devices.

Summary of changes:
  * Adds new execution module `salt/modules/nxos_upgrade.py`
  * Adds new state module `salt/states/nxos_upgrade.py`
  * Adds new `NxosError` exception classes

### Tests written?

Manual tests executed on the following platform types.

* N9k (disruptive and non-disruptive upgrades)
    * Tested both traditional ISSU (Dual Supervisors) and Enhanced ISSU (LXC boot mode)
    * SSH and NX-API Proxy Minion
    * Native GuestShell Minion
* N7k
    * SSH and NX-API Proxy Minion
* N5k
    * SSH Proxy Minion
* N3k
    * SSH and NX-API Proxy Minion
    * Native GuestShell Minion

### Commits signed with GPG?

Yes

### CODE REVIEW

File: salt/modules/nxos_upgrade.py [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+    system_image (Mandatory Option)
+        Path on bootflash: to system image upgrade file.
+
+    kickstart_image
+        Path on bootflash: to kickstart image upgrade file.
+        (Not required if using combined system/kickstart image file)
+        Default: None
+
+    issu
+        Set this option to True when an in service software service or
```
Comment (tstoner) [inline salt/modules/nxos_upgrade.py]: did you mean? In Service Software Upgrade

Comment (mikewiebe) [reply salt/modules/nxos_upgrade.py]: Yes.  Good catch.

File: salt/exceptions.py [hunk 2]
```diff
@@ -566,3 +566,27 @@ class MissingSmb(SaltException):
...
+
+class NxosClientError(NxosError):
+    '''
+    NX-OS Client Error raised for problems connecting to the NX-OS device
+    '''
+
+
+class NxosRequestNotSupported(NxosError):
+    '''
+    Raised for unssuported client requests
```
Comment (cachedout) [inline salt/exceptions.py]: `unsupported`

File: salt/utils/nxos.py [hunk 3]
```diff
@@ -1,4 +1,18 @@
 # -*- coding: utf-8 -*-
+# Copyright (c) 2018 Cisco and/or its affiliates.
```
Comment (cachedout) [inline salt/utils/nxos.py]: Is this line necessary?

Comment (mikewiebe) [reply salt/utils/nxos.py]: This is pretty standard for contributions Cisco makes to open source projects.  Why do you ask?

File: salt/exceptions.py (hunk 2 above)
Comment (mikewiebe) [reply salt/exceptions.py]: Will fix.

File: salt/utils/nxos.py (hunk 3 above)
Comment (cachedout) [reply salt/utils/nxos.py]: It's not a problem. I'll resolve this review comment.

File: salt/exceptions.py (hunk 2 above)
Comment (cachedout) [reply salt/exceptions.py]: I just went ahead and pushed a quick fix.

**Comment by mikewiebe** [comment]
@cachedout Thanks for fixing my spelling error and the quick review!

Couple of questions:

* I was not sure what to do for the version `versionadded:: xxxx.xx.x`
* I was able to solve all of the pylint issues except for 2 that indicated python3 support issues.  There were no additional details however so I am not exactly sure what the problem is and I don't see the report now.

**Comment by mikewiebe** [comment]
@cachedout Just checking on this PR.  I posted a few questions above.  Other then that is there anything that is pending before we can merge this?  Thanks!

**Comment by cachedout** [comment]
@mikewiebe We'll update the versionadded for you at release time. I went ahead and fixed the lint issues for you also. Once tests finish, this should be ready to merge.

**Comment by rallytime** [comment]
@mikewiebe This is causing some tests to fail: https://jenkinsci.saltstack.com/job/pr-kitchen-centos7-py2/job/PR-50395/12/

Can you take a look?

**Comment by mikewiebe** [comment]
@rallytime I am having trouble determining how the integration test failures are related to my changes.  

I picked the first test from the list of failures and ran it in my dev environment.

**Test:** `integration.proxy.test_shell.ProxyCallerSimpleTestCase.test_can_it_ping`

When I first ran the test in my `nxos_upgrade` branch I see the same failure reported...

.... but keep reading ...

```shell
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Python Version: 2.7.13 (default, Apr 4 2017, 08:47:57) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)]
 * Transplanting configuration files to '/private/tmp/salt-tests-tmpdir/config'
 * Current Directory: /Users/mwiebe/Projects/salt
 * Test suite is running under PID 94392
 * Logging tests on /private/tmp/salt-runtests.log
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Max open files settings is too low (soft: 256, hard: 9223372036854775807) for running the tests
 * Trying to raise the limits to soft: 3072, hard: 9223372036854775807
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Setting up Salt daemons to execute tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Starting salt-master ... STARTED!                                                                                                                                                             
 * Starting salt-minion ... STARTED!                                                                                                                                                             
 * Starting sub salt-minion ... STARTED!                                                                                                                                                         
 * Starting syndic salt-master ... STARTED!                                                                                                                                                      
 * Starting salt-syndic ... STARTED!                                                                                                                                                             
=================================================================================================================================================================================================
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Starting integration.proxy.test_shell.ProxyCallerSimpleTestCase.test_can_it_ping Tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Could not access /private/tmp/salt-tests-tmpdir/rootdir-proxy/pki. Path does not exist.
E
======================================================================
ERROR: test_can_it_ping (integration.proxy.test_shell.ProxyCallerSimpleTestCase)
[CPU:64.2%|MEM:69.5%] 
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/mwiebe/Projects/salt/tests/integration/proxy/test_shell.py", line 31, in test_can_it_ping
    ret = self._load_return(self.run_call('--proxyid proxytest --out=json test.ping'))
  File "/Users/mwiebe/Projects/salt/tests/integration/proxy/test_shell.py", line 25, in _load_return
    return json.loads('\n'.join(ret))
  File "/Users/mwiebe/Projects/salt/salt/utils/json.py", line 92, in loads
    return json_module.loads(s, **kwargs)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 382, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded

----------------------------------------------------------------------
Ran 1 test in 1.659s

FAILED (errors=1)
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 231, in serve_forever
    poll_interval)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 150, in _eintr_retry
    return func(*args)
error: (9, 'Bad file descriptor')


====================================================================================  Overall Tests Report  =====================================================================================
*** integration.proxy.test_shell.ProxyCallerSimpleTestCase.test_can_it_ping Tests  **************************************************************************************************************
 --------  Tests with Errors  -------------------------------------------------------------------------------------------------------------------------------------------------------------------
   -> integration.proxy.test_shell.ProxyCallerSimpleTestCase.test_can_it_ping  ..................................................................................................................
       Traceback (most recent call last):
         File "/Users/mwiebe/Projects/salt/tests/integration/proxy/test_shell.py", line 31, in test_can_it_ping
           ret = self._load_return(self.run_call('--proxyid proxytest --out=json test.ping'))
         File "/Users/mwiebe/Projects/salt/tests/integration/proxy/test_shell.py", line 25, in _load_return
           return json.loads('\n'.join(ret))
         File "/Users/mwiebe/Projects/salt/salt/utils/json.py", line 92, in loads
           return json_module.loads(s, **kwargs)
         File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 339, in loads
           return _default_decoder.decode(s)
         File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 364, in decode
           obj, end = self.raw_decode(s, idx=_w(s, 0).end())
         File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 382, in raw_decode
           raise ValueError("No JSON object could be decoded")
       ValueError: No JSON object could be decoded
   ..............................................................................................................................................................................................
 ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
=================================================================================================================================================================================================
FAILED (total=1, skipped=0, passed=0, failures=0, errors=1) 
====================================================================================  Overall Tests Report  =====================================================================================
```

I noticed the following message printed at the very beginning of the test run.

`Could not access /private/tmp/salt-tests-tmpdir/rootdir-proxy/pki. Path does not exist.`

I decided to create the directory and re-run the test.  With the directory in place the test passes.

```shell
(py2-salt) saltstack_master:~/Projects/salt$mkdir -p /private/tmp/salt-tests-tmpdir/rootdir-proxy/pki
(py2-salt) saltstack_master:~/Projects/salt$
```

Now with the re-run it passes.

```shell
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Python Version: 2.7.13 (default, Apr 4 2017, 08:47:57) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)]
 * Transplanting configuration files to '/private/tmp/salt-tests-tmpdir/config'
 * Current Directory: /Users/mwiebe/Projects/salt
 * Test suite is running under PID 96611
 * Logging tests on /private/tmp/salt-runtests.log
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Max open files settings is too low (soft: 256, hard: 9223372036854775807) for running the tests
 * Trying to raise the limits to soft: 3072, hard: 9223372036854775807
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Setting up Salt daemons to execute tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Starting salt-master ... STARTED!                                                                                                                                                             
 * Starting salt-minion ... STARTED!                                                                                                                                                             
 * Starting sub salt-minion ... STARTED!                                                                                                                                                         
 * Starting syndic salt-master ... STARTED!                                                                                                                                                      
 * Starting salt-syndic ... STARTED!                                                                                                                                                             
=================================================================================================================================================================================================
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Starting integration.proxy.test_shell.ProxyCallerSimpleTestCase.test_can_it_ping Tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[WARNING ] top_file_merging_strategy is set to 'merge' and multiple top files were found. Merging order is not deterministic, it may be desirable to either set top_file_merging_strategy to 'same' or use the 'env_order' configuration parameter to specify the merging order.
.
----------------------------------------------------------------------
Ran 1 test in 13.743s

OK
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 231, in serve_forever
    poll_interval)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 150, in _eintr_retry
    return func(*args)
error: (9, 'Bad file descriptor')


====================================================================================  Overall Tests Report  =====================================================================================
***  No Problems Found While Running Tests  *****************************************************************************************************************************************************
=================================================================================================================================================================================================
OK (total=1, skipped=0, passed=1, failures=0, errors=0) 
====================================================================================  Overall Tests Report  =====================================================================================
```

I still see the `Bad file descriptor` traceback but the Test report says it passed and the `ValueError: No JSON object could be decoded` is gone.

I see the same exact behavior running in the `develop` branch without my changes.

Not sure where to go from here....

**Comment by mikewiebe** [comment]
@rallytime I rebased but still see the PR is reporting the integration test failures.  I re-ran two of them locally and the both pass for me.  After the rebase I don't have to manually create the directory.

```shell
(py2-salt) mwiebe@MWIEBE-M-D2TB:~/Projects/salt$./tests/runtests.py --name=integration.proxy.test_shell.ProxyCallerSimpleTestCase.test_can_it_ping
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Python Version: 2.7.13 (default, Apr 4 2017, 08:47:57) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)]
 * Transplanting configuration files to '/private/tmp/salt-tests-tmpdir/config'
 * Current Directory: /Users/mwiebe/Projects/salt
 * Test suite is running under PID 29560
 * Logging tests on /private/tmp/salt-runtests.log
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Max open files settings is too low (soft: 256, hard: 9223372036854775807) for running the tests
 * Trying to raise the limits to soft: 3072, hard: 9223372036854775807
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Setting up Salt daemons to execute tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Starting salt-master ... STARTED!                                                                                                                                                                      
 * Starting salt-minion ... STARTED!                                                                                                                                                                      
 * Starting sub salt-minion ... STARTED!                                                                                                                                                                  
 * Starting syndic salt-master ... STARTED!                                                                                                                                                               
 * Starting salt-syndic ... STARTED!                                                                                                                                                                      
==========================================================================================================================================================================================================
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Starting integration.proxy.test_shell.ProxyCallerSimpleTestCase.test_can_it_ping Tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[WARNING ] top_file_merging_strategy is set to 'merge' and multiple top files were found. Merging order is not deterministic, it may be desirable to either set top_file_merging_strategy to 'same' or use the 'env_order' configuration parameter to specify the merging order.
Exception in thread Thread-1 (most likely raised during interpreter shutdown):
.
----------------------------------------------------------------------
Ran 1 test in 14.366s

OK
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 231, in serve_forever
    poll_interval)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 150, in _eintr_retry
    return func(*args)
error: (9, 'Bad file descriptor')


=========================================================================================  Overall Tests Report  =========================================================================================
***  No Problems Found While Running Tests  **************************************************************************************************************************************************************
==========================================================================================================================================================================================================
OK (total=1, skipped=0, passed=1, failures=0, errors=0) 
=========================================================================================  Overall Tests Report  =========================================================================================
(py2-salt) mwiebe@MWIEBE-M-D2TB:~/Projects/salt$./tests/runtests.py --name=integration.proxy.test_shell.ProxyCallerSimpleTestCase.test_grains_items
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Python Version: 2.7.13 (default, Apr 4 2017, 08:47:57) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)]
 * Transplanting configuration files to '/private/tmp/salt-tests-tmpdir/config'
 * Current Directory: /Users/mwiebe/Projects/salt
 * Test suite is running under PID 30750
 * Logging tests on /private/tmp/salt-runtests.log
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Max open files settings is too low (soft: 256, hard: 9223372036854775807) for running the tests
 * Trying to raise the limits to soft: 3072, hard: 9223372036854775807
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Setting up Salt daemons to execute tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Starting salt-master ... STARTED!                                                                                                                                                                      
 * Starting salt-minion ... STARTED!                                                                                                                                                                      
 * Starting sub salt-minion ... STARTED!                                                                                                                                                                  
 * Starting syndic salt-master ... STARTED!                                                                                                                                                               
 * Starting salt-syndic ... STARTED!                                                                                                                                                                      
==========================================================================================================================================================================================================
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Starting integration.proxy.test_shell.ProxyCallerSimpleTestCase.test_grains_items Tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[WARNING ] top_file_merging_strategy is set to 'merge' and multiple top files were found. Merging order is not deterministic, it may be desirable to either set top_file_merging_strategy to 'same' or use the 'env_order' configuration parameter to specify the merging order.
.
----------------------------------------------------------------------
Ran 1 test in 13.391s

OK
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 231, in serve_forever
    poll_interval)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 150, in _eintr_retry
    return func(*args)
error: (9, 'Bad file descriptor')


=========================================================================================  Overall Tests Report  =========================================================================================
***  No Problems Found While Running Tests  **************************************************************************************************************************************************************
==========================================================================================================================================================================================================
OK (total=1, skipped=0, passed=1, failures=0, errors=0) 
=========================================================================================  Overall Tests Report  =========================================================================================
(py2-salt) mwiebe@MWIEBE-M-D2TB:~/Projects/salt$
```

**Comment by cachedout** [comment]
@mikewiebe The quickest way to remedy it would just be to close this PR and open another one. :)

**Comment by mikewiebe** [comment]
@cachedout I am glad you confirmed that because I was literally in the process of doing just that :).

**Comment by mikewiebe** [comment]
PR https://github.com/saltstack/salt/pull/50588 replaces this PR.

**Pull request closed without merging**